code/ai/aicode.cpp

```diff
@@ -6159,6 +6168,57 @@ int check_ok_to_fire(int objnum, int target_objnum, weapon_info *wip)
 	return 1;
 }
 
+//	--------------------------------------------------------------------------
+//  Returns true if *aip has a line of sight to its current target.
+//	This is a computationally expensive operation, so use sparingly
+int check_los(int objnum, int target_objnum) {
+	vec3d& end = Objects[target_objnum].pos;
+	vec3d& start = Objects[objnum].pos;
+
+	for (int i = 0; i < Highest_object_index; i++) {
+		//Don't collision check against ourselves or our target
+		if (i == objnum || i == target_objnum)
+			continue;
+
+		int model_num = 0;
+		int model_instance_num = 0;
+
+		//Only collision check against other pieces of Debris, Asteroids or Ships
+		char type = Objects[i].type;
+		if (type == OBJ_DEBRIS) {
+			model_num = Debris[Objects[i].instance].model_num;
+			model_instance_num = -1;
+		}
+		else if (type == OBJ_ASTEROID) {
+			model_num = Asteroid_info[Asteroids[Objects[i].instance].asteroid_type].model_num[Asteroids[Objects[i].instance].asteroid_subtype];
+			model_instance_num = Asteroids[Objects[i].instance].model_instance_num;
+		}
+		else if (type == OBJ_SHIP) {
+			model_num = Ship_info[Ships[Objects[i].instance].ship_info_index].model_num;
+			model_instance_num = Ships[Objects[i].instance].model_instance_num;
+		}
+		else
+			continue;
+
+		mc_info hull_check;
+		mc_info_init(&hull_check);
+
+		hull_check.model_instance_num = model_instance_num;
+		hull_check.model_num = model_num;
+		hull_check.orient = &Objects[i].orient;
+		hull_check.pos = &Objects[i].pos;
+		hull_check.p0 = &start;
+		hull_check.p1 = &end;
+		hull_check.flags = MC_CHECK_MODEL;
+
+		if (model_collide(&hull_check)) {
+			return 0;
+		}
+	}
+
+	return 1;
```

Comment:
Firstly, since we have the benefit of C++, this can return Boolean instead of an int.

As you say, a model checking collision isn't exactly cheap so we want to avoid it where possible using simpler data. For instance if neither us nor our target is inside our possibly-line-of-sight-blocking object and it is farther away than our target plus its radius we know it couldn't possibly occlude us. It could also be in the wrong direction as our target, there's a few of these so I might directly code suggest them a little bit later.

Furthermore its important to consider what the intended use of this function really is. You probably don't want to hold your fire if a fighter or fast moving peice of debris happens to be in the way, its likely that our shot won't even hit them by time it gets to them because they'll be long gone by then, so I think it might be preferable to check for small or fast-moving objects and just immediately disregard them saving a lot of computation. Or you could have an optional parameter for it if you really want strict LoS.

Reply:
> distance less than its radius
I did think about that, but as model_collide() has a pretty quick exit with a Bounding Sphere check, I assumed that would be redundant here
> if a fighter or fast moving peice of debris
it would be an option, I will be adding minRadius as a parameter
> maybe_avoid_big_ship´ could then use this function instead
I can see what I can do

Reply:
You should check out #2735 . Once I started profiling these things the results were quite surprising. Doing this checking here instead of just letting ´segment_sphere´ handle it in ´model_collide´ is a pretty miniscule time saving but this early outing will be happening _a lot_, it will be the overwhelmingly common situation and suddenly that time saving becomes not so miniscule, because simply setting up the model collision takes a significant portion of that time.

That said, I think only 3 simple checks are really necessary here, whether you or the target are inside the objects radius (no avoiding the model collision then) checking that the object is between you and the target, and that its close enough to the path to occlude it. And I can help you with those in a few hours if you're not sure how to implement them.

Reply:
Okay, added some earlier outs for small radius (for now I assume things with a radius smaller than 1 are irrelevant for los when firing weapons), as well as not relevant for los due to distance from the relevant los as well as being behind target or ship.
Regarding usage in maybe_avoid_big_ship, I'd like to make that a separate PR, as that needs in itself a lot more testing as you'd need to make sure you're not breaking anything that relies on this AI behaviour

Reply:
And while I didn't expect there to be that much difference, your profiling results speak for themselves.